### ISSUE TITLE
incorrect `InvalidTemplateParam` using when a `Closure` argument uses `T` as an argument.

### ISSUE THREAD

**Issue by azjezz** [vimeo/psalm#8212]
given the following example: https://psalm.dev/r/7f1d36d1f8, the error here is incorrect.

`Option` type is immutable by itself, and adding `@psalm-immutable` annotation solves the problem: https://psalm.dev/r/3deed330bf

however, this leads to another issue, at this point, `Option<T>::map<Tu>((Closure(T): Tu) $f): Option<Tu>` is considered pure, allowing to declare `get_organization_owner()` function below as pure ( see: https://psalm.dev/r/4396258d6f ), this leads to another problem: `get_organization_owner()` is not actually pure, neither is `map()` function, `get_organization_owner()` calls `Option::map()`, which calls `$f`, which is an impure function ( in the example above it is indeed pure, but `Organization::getOwner()` can have side effects ), declaring just `Option::map()` as impure based on the fact that it calls an impure function ( `$f` ) that could result in side effects elsewhere but not in the option object itself is not possible, meaning we are forced to remove the `@psalm-immutable` annotation from the class `Option`, leading us back to the first issue.

**Comment by azjezz** [comment]
I think the issue boils down to this: https://psalm.dev/r/eb422e250d

not allowing an immutable object to call an impure function, even though we know for a fact that it cannot change the state of the object.

maybe a new annotation `@psalm-internal-mutation-free` annotation can be used in such a function to declare that it is not totally pure, but it won't change the object itself, allowing the use of it in `@psalm-immutable` classes only.

**Comment by azjezz** [comment]
for example, there's no difference technically between: https://psalm.dev/r/eb422e250d and https://psalm.dev/r/7be3725e19, but one is valid, while the other is not.

**Comment by AndrolGenhald** [comment]
Seems like the same sort of problem as #8116.

**Comment by azjezz** [comment]
I think i have a better solution for this problem.

The `Option` class given in the first example, is not immutable ( by psalm definition, which is completely side-effect free ), but rather a readonly class, what we can do is add support for `@psalm-readonly` annotation on classes ( same as `readonly class Foo { .. }` in PHP 8.2 ), so we would allow `Closure` arguments to take `T` as an argument themselves, and we know that the class is not side effect free, it's just that it's own state doesn't change.

I'll send a PR with tests which might help you understand what i mean more ^^"

**Comment by someniatko** [comment]
> I think the issue boils down to this: https://psalm.dev/r/eb422e250d
> 
> not allowing an immutable object to call an impure function, even though we know for a fact that it cannot change the state of the object.

`$this->value` is passed by reference to `$f`. Therefore, it may be mutated. I think Psalm behaves correctly, and it has to accept something like `pure-callable` but for Closure.

**Comment by azjezz** [comment]
No, it actually has nothing to do with objects: https://psalm.dev/r/3033cae1ca

**Comment by someniatko** [comment]
Closure could still take its argument by reference though

**Comment by AndrolGenhald** [comment]
@someniatko huh, [you're right](https://psalm.dev/r/e59d68ba57). I'd consider that a separate bug though. I think we need to add annotation for references in callables and make `callable(&int)` not be a subtype of `callable(int)`, because the example I just gave [crashes](https://3v4l.org/T3J6J)...
